packages/doenetml-prototype/src/state/redux-slices/dast/utils/normalize-dast.ts

```diff
@@ -0,0 +1,146 @@
+import { Plugin } from "unified";
+import {
+    DastElement,
+    DastElementContent,
+    DastRoot,
+    DastRootContent,
+    visit,
+} from "@doenet/parser";
+import { unified } from "unified";
+
+/**
+ * Normalize the DAST tree so that it is contained in a single `<document>` element.
+ * As well, remove commends and XML-valid but not-useful-for-DoenetML nodes like XML instructions and doctypes.
+ */
+export function normalizeDocumentDast(dast: DastRoot) {
+    const processor = unified()
+        .use(pluginRemoveComments)
+        .use(pluginChangeCdataToText)
+        .use(pluginEnsureDocumentElement);
+    return processor.runSync(dast);
+}
+
+/**
+ * Change all CDATA nodes to text nodes.
+ */
+const pluginChangeCdataToText: Plugin<[], DastRoot, DastRoot> = () => {
+    return (tree) => {
+        visit(tree, (node) => {
+            if (node.type === "cdata") {
+                // @ts-ignore
+                node.type = "text";
+            }
+        });
+    };
+};
+
+/**
+ * Remove all comment nodes from the DAST tree.
+ */
+const pluginRemoveComments: Plugin<[], DastRoot, DastRoot> = () => {
+    return (tree) => {
+        visit(tree, (node) => {
+            if (node.type === "element" || node.type === "root") {
+                node.children = node.children.filter(
+                    (n) => n.type !== "comment",
+                );
+            }
+        });
+    };
+};
+
+/**
+ * Plugin that ensure the root of the document is a single `<document>` element
+ * and that all whitespace is trimmed from the content of `<document>`.
+ */
+const pluginEnsureDocumentElement: Plugin<[], DastRoot, DastRoot> = () => {
+    return (tree) => {
+        const existingDocument = tree.children.find(
+            (n) => n.type === "element" && n.name === "document",
+        ) as DastElement | undefined;
+        if (existingDocument) {
+            existingDocument.children = trimWhitespaceNodes(
```

Comment:
filterNonDoenetXml here too?

Reply:
Ah, I didn't do that because docstrings are only valid at the root of the document. Since we're throwing away the rest of the root, it should be okay.